### ISSUE TITLE
Adding Resource doesn't work in WinUI 3 for desktop

### ISSUE THREAD

**Issue by arivoir** [microsoft/microsoft-ui-xaml#4114]
When adding a xaml resource the project doesn't compile

```
<Grid.Resources>
     <local:MyClass x:Key="asdf"/>
</Grid.Resources>
```

> Error	WMC0026	Cannot add 'MyClass' into the dictionary property 'Resources', type must be 'Object'	C:\Projects\XXXXXX\MainWindow.xaml	48	

Additionally the Window object has no Resource property.

**Comment by marb2000** [comment]
Hi @arivoir, as you pointed out, the Window object doesn't have Resource property like WPF, for instance. It was a design decision, however, you are welcome to open a feature request.  Also, can you share your project, or give me a the complete repro steps so I can understand the scenario? It's hard to me to help you with so little info. Thanks!

**Comment by arivoir** [comment]
I can leave without the Window's resource dictionary. Anyway there seems to be a bug when adding a resource in a framework element like Grid.

I could add a style

```
        <Grid.Resources>
            <Style TargetType="TextBlock" x:Key="titleStyle">
                <Setter Property="FontSize" Value="{ThemeResource HubHeaderThemeFontSize}"/>
                <Setter Property="FontWeight" Value="{ThemeResource HubHeaderThemeFontWeight}"/>
                <Setter Property="Margin" Value="20 30 20 5" />
            </Style>
        </Grid.Resources>
```

But when the object is not a style, the project doesn't compile. Should the resource item inherit from something specific? The error says it must be 'Object', but clearly the message is not true.

**Comment by marb2000** [comment]
@arivoir I can't repro this issue with the latest bits of the Preview 4 that we will ship soon. Perhpahs they were an issue within the preview 3. This is the code that I used on my Window:

```XML
 <Grid>
        <Grid.Resources>
            <Style TargetType="TextBlock" x:Key="titleStyle">
                <Setter Property="FontSize" Value="{ThemeResource HubHeaderThemeFontSize}"/>
                <Setter Property="FontWeight" Value="{ThemeResource HubHeaderThemeFontWeight}"/>
                <Setter Property="Margin" Value="20 30 20 5" />
            </Style>
        </Grid.Resources>
        <StackPanel Orientation="Horizontal" HorizontalAlignment="Center" VerticalAlignment="Center">
            <Button x:Name="myButton" Click="myButton_Click">Click Me</Button>
            <TextBlock x:Name="myText" Style="{ThemeResource titleStyle}" ></TextBlock>
        </StackPanel>
    </Grid>
```
If you want, we can keep this bug open until you try with the new bits. Sounds good?

**Comment by arivoir** [comment]
Please try creating whatever class, not inheriting from anything(inheriting from object) and see whether you can add it as a resource.

**Comment by marb2000** [comment]
Can you attack a sample of your scenario please?

**Comment by arivoir** [comment]
The reproduction is very simple.

1. Create a new project
2. Add a new class 
![image](https://user-images.githubusercontent.com/17496741/107371851-81931880-6ac3-11eb-9dcf-d61bc8d27516.png)

3. Add that class as a resource in xaml 
![image](https://user-images.githubusercontent.com/17496741/107372006-b606d480-6ac3-11eb-9341-1c2433645b08.png)

> Severity	Code	Description	Project	File	Line	Suppression State
> Error	WMC0026	Cannot add 'MyClass' into the dictionary property 'Resources', type must be 'Object'	App2	App2\App2\MainWindow.xaml	12	

I attached the repro project

[App2.zip](https://github.com/microsoft/microsoft-ui-xaml/files/5951698/App2.zip)

**Comment by marb2000** [comment]
Got it! Thanks! I think it should say a DependencyObject, @ranjeshj do you know the details?

**Comment by ranjeshj** [comment]
I don't think it needs to be DependencyObject in order to be placed in the resource dictionary. If it takes object, that should just work. Perhaps the markup compiler is getting tripped by something. @stevenbrix @alwu-msft any ideas ?

**Comment by duke7553** [comment]
@marb2000 Seems like I get the same error on build in WinUI Desktop Preview 4 for specific resources in my App.xaml:
![image](https://user-images.githubusercontent.com/20365014/107894497-cbce2c80-6efd-11eb-9c30-74b053335d80.png)

It occasionally helps to clean, rebuild, and cross your fingers. But I've found this to be unreliable.

**Comment by marb2000** [comment]
Pinging @stevenbrix and @alwu-msft again.

**Comment by alwu-msft** [comment]
@RealTommyKlein this is possibly a regression in XamlCompiler as this scenario works in a UWP XAML app.

**Comment by RealTommyKlein** [comment]
Thanks for the report, this issue is specific to C# Desktop apps using .NET 5.  It'll be fixed in the next release of WinUI 3.

**Comment by billhenn** [comment]
@RealTommyKlein - Do you have a ballpark idea of when the next release with the fix will be out?  I am just getting into WinUI 3, trying to port things from UWP and am running into this issue in multiple spots.  Any chance it will be out soon?

**Comment by RealTommyKlein** [comment]
@billhenn - we haven't announced a date for the next release, but the monthly WinUI community calls are the best source of information.  The [next one on March 17th](https://www.youtube.com/watch?v=gnPJ0t7rFnA&ab_channel=WindowsDeveloper) should have more concrete information on our next release.

**Comment by RealTommyKlein** [comment]
Marking as closed since this should be resolved in the `0.5-prerelease` version of Project Reunion.
